File: cipher/hill_cypher.cpp
// Name: Kartikey Tewari
// Roll No.: 17MI551
// Computer Science and Engineering (Dual Degree)
// National Institute of Technology, Hamirpur
#include <bits/stdc++.h>
using namespace std;

string encrypt (string s, string password)
{
    int n=s.length();
    int plain_text[n];
    for (int i=0;i<=n-1;i++)
    {
        plain_text[i]=int(s[i])-int('a');
    }

    int key[n][n];
    for (int i=0;i<=n-1;i++)
    {
        for (int j=0;j<=n-1;j++)
        {
            int location=(n*i)+j;
            key[i][j]=int(password[location])-int('a');
        }
    }

    // (key*plain_text)%26=cipher_text
    int cipher_text[n];
    for (int i=0;i<=n-1;i++)
    {
        cipher_text[i]=0;
        for (int j=0;j<=n-1;j++)
        {
            cipher_text[i]+=key[i][j]*plain_text[j];
        }
        cipher_text[i]%=26;
    }

    string ans="";
    for (int i=0;i<=n-1;i++)
    {
        cout << "cipher_text= " << cipher_text[i] << endl;
        ans+=char(cipher_text[i]+97);
    }

    return ans;
}

string decrypt (string s, string password)
{
    int n=s.length();
    int cipher_text[n];
    for (int i=0;i<=n-1;i++)
    {
        cipher_text[i]=int(s[i])-int('a');
    }

    int key[n][n];
    for (int i=0;i<=n-1;i++)
    {
        for (int j=0;j<=n-1;j++)
        {
            int location=(i*n)+j;
            key[i][j]=int(s[location])-int('a');
        }
    }

    int inverse_key[n][n];
    // generate inverse_key 

    int plain_text[n];
    for (int i=0;i<=n-1;i++)
    {
        plain_text[i]=0;
        for (int j=0;j<=n-1;j++)
        {
            plain_text[i]+=inverse_key[i][j]*cipher_text[j];
        }
        plain_text[i]%=26;
    }

    string ans="";
    for (int i=0;i<=n-1;i++)
    {
        ans+=char(plain_text[i]+97);
    }
    return ans;
}

int main()
{
    int enc_bit;
    cout << "Enter 1 for encryption" << endl;
    cout << "Enter 2 for decryption" << endl << endl;
    cin >> enc_bit;

    if (enc_bit==1)
    {
        // encryption
        string s, password;
        cout << "Enter string for enryption: ";
        cin >> s;
    
        cout << "Enter a password: " << endl;
        cout << "Password length should be " << s.length()*3 << " characters: ";
        cin >> password;
        if (password.length()!=3*s.length())
        {
            cout << "Incorrect password length. System terminating." << endl;
        }
        else
        {
            cout << "Encrypted string: " << encrypt(s,password) << endl;
        }
    }
    else
    {
        // decrption
        string s, password;
        cout << "Enter the string to decrypt= ";
        cin >> s;
        cout << "Enter the password= ";
        cin >> password;

        if (password.length()!=3*s.length())
        {
            cout << "Incorrect password length. System Terminating." << endl;
        }
        else
        {
            cout << "Decrypted String: " << decrypt(s,password) << endl;
        }
    }

    return 0;
}

File: cipher/hill_cypher.py
# Name: Kartikey Tewari
# Roll No.: 17MI551
# Computer Science and Engineering (Dual Degree)
# National Institute of Technology, Hamirpur
import numpy as np

def encrypt (s, password):
    n=len(s)
    
    plain_text=np.arange(n).reshape(n,1)
    for i in range(0,n):
        plain_text[i][0]=ord(s[i])-ord('a')

    key=np.arange(n*n).reshape(n,n)
    for i in range(0,n):
        for j in range(0,n):
            key[i][j]=ord(password[(n*i)+j])-ord('a')

    encrypted_text=key.dot(plain_text)

    ans=""
    for i in range(0,n):
        # print(encrypted_text[i])
        ans+=chr((encrypted_text[i][0]%26)+ord('a'))
    
    return ans

def decrypt (s, password):
    n=len(s)

    encrypted_text=np.arange(n).reshape(n,1)
    for i in range(0,n):
        encrypted_text[i][0]=ord(s[i])-ord('a')

    key=np.arange(n*n).reshape(n,n)
    for i in range(0,n):
        for j in range(0,n):
            key[i][j]=ord(password[(n*i)+j])-ord('a')


    inverse_key=np.linalg.inv(np.matrix(key))
    plain_text=inverse_key.dot(encrypted_text)

    ans=""
    for i in range(0,n):
        ans+=chr((plain_text[i][0]%26)+ord('a'))

    return ans

print("Enter 1 for encryption")
print("Enter 2 for decryption")
a=int(input())

if (a==1):
    # encryption
    print("Enter string for encrypton:")
    s=str(input())
    n=len(s)
    print("Password should be " + str(n*n) + " letters long")
    print("Enter password: ")
    password=str(input())
    if (len(password)!=n*n):
        print("Programme terminating. Invalid length of string/password.")
    else:
        encrypted_string=encrypt(s,password)
        print("The encrypted string is: " + encrypted_string)

if (a==2):
    # decryption
    print("Enter string for decryption:")
    s=str(input())
    n=len(s)
    print("Enter password: ")
    password=str(input())
    if (len(password)!=n*n):
        print("Process termination. Invalid length of string.")
    else:
        decrypted_string=decrypt(s,password)
        print("The decrypted string is " + decrypted_string)

